Start on Layouts for Phones

hatton opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1107)
<!-- Reviewable:end -->

JohnThomson commented:
Wonder whether we're getting enough cases of "find an htm or html" file or "list the htm{l} files" to be worth a method somewhere? e.g., TryGetHtmlFile(string path, out whereFound), returns true if changing path's extension to either htm or html finds a file.

---

Reviewed 389 of 389 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/BloomBrowserUI/gulpfile.js, line 23 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKE77BtmPgnqvq-xOa4:-KKE77Bu-VAeGF-P9xF6:765389691) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomBrowserUI/gulpfile.js#L23)):*
> ```JavaScript
> var paths = {
>    less: ['./**/*.less',  '!./node_modules/**/*.less','!./output/**/*.*'],
>    jade: ['./**/*.jade',  '!./node_modules/**/*.jade','!./**/*mixins.jade','!./output/**/*.*'],
> ```

If we ought to remove these (as the todo above says) why are we adding to them?

---

*[src/BloomBrowserUI/gulpfile.js, line 27 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKE7GfqwaO4Yq16u72M:-KKE7GfqwaO4Yq16u72N:-1390280835) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomBrowserUI/gulpfile.js#L27)):*
> ```JavaScript
>    
>    //files we are *not* running through some compiler that need to make it into the outputDir directory.
>    filesThatMightBeNeededInOutput: ['./**/*.*', '!./**/*.ts','!./**/*.jade','!./**/*.less','!./**/*.bat','!./**/node_modules/**/*.*','!./output/**/*.*'],
> ```

But surely we DO run jade and less files through compilation? Maybe there are now SOME that we need, but surely not all of them? At least the comment should change.

---

*[src/BloomBrowserUI/templates/template books/Decodable Reader/Basic Book.css, line 1 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKEBBuyPsqWXFIaFbqs:-KKEBBuzxiFP3nWvad8Y:1925484739) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomBrowserUI/templates/template books/Decodable Reader/Basic Book.css#L1)):*
> ```CSS
> /*+}*/
> ```

Is it intentional to check in this css file? Usually now we just check in the .less. (I assume most or all of this isn't really new so I haven't tried to review it.)

---

*[src/BloomExe/BloomFileLocator.cs, line 178 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKECyLeupkKLLyoeerH:-KKECyLeupkKLLyoeerI:-117624023) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomExe/BloomFileLocator.cs#L178)):*
> ```Smalltalk
> 		public static string GetFactoryBookTemplateDirectory(params string[] parts)
> 		{
> 			return Path.Combine(FactoryTemplateBookDirectory, parts[0]);
> ```

Feels odd. Why take multiple arguments and only use one? Needs comment at least.

---

*[src/BloomExe/Book/BookStorage.cs, line 590 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKESuRjE3wESXvzkHJc:-KKESuRjE3wESXvzkHJd:1242117288) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomExe/Book/BookStorage.cs#L590)):*
> ```Smalltalk
> 			if (File.Exists(p))
> 				return p;
> 			p = Path.Combine(folderPath, "templatePages.html");
> ```

I realize all you did here was add the html option, but the overall logic escapes me. Why is it OK to have multiple HTML files in a book folder as long as one of them is called templatePages? Why would we assume that is the book? Our factory templates are called e.g. Basic Book.htm, not templatePages.html.

---

*[src/BloomExe/Book/Layout.cs, line 80 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKEU9XsO9XK_9zfhDAx:-KKEU9XsO9XK_9zfhDAy:-1759471519) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomExe/Book/Layout.cs#L80)):*
> ```Smalltalk
> 		}
> 
> 		public bool IsDeviceLayout
> ```

I think it might be helpful to say something here about the significance of a layout being a 'device' layout. I guess it means something like, a layout intended for a (class of) devices rather than for paper (or PDF). But it might be helpful to describe what difference it makes and why. I suppose one can look up usages, but that doesn't always make clear the _intent_: why do we need to distinguish layouts intended for devices? Just so we can refuse to print a device layout? Or is there more to it?

Not a problem if you decide it's obvious using Usages. I don't have that convenience in a review.

---

*[src/BloomExe/Collection/BookCollection.cs, line 96 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKEVTSUcX52Z7LAhEGu:-KKEVTSUcX52Z7LAhEGv:-432504001) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomExe/Collection/BookCollection.cs#L96)):*
> ```Smalltalk
> 			{
> 				var dirName = Path.GetFileName(_path);
> 				return dirName == "template books" ? "Templates" : dirName;
> ```

Why do we want to use a name for the folder different from the name we show the user?

Should we localize any of the collection names? (Assuming this Name is displayed to the user.)

---

*[src/BloomExe/MiscUI/BloomIntegrityDialog.cs, line 31 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKEWYA-wbRZLZ5f0eSG:-KKEWYA0GL_BounOGOS5:-1611441193) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/eb8c9f3d3b29a842e6bc62f95e316969dc16e207/src/BloomExe/MiscUI/BloomIntegrityDialog.cs#L31)):*
> ```Smalltalk
> 			string[] dirs;
> 			if (SIL.PlatformUtilities.Platform.IsWindows)
> 				dirs = new[] { "AndikaNewBasic", "factoryCollections", "localization", "xslts" };
> ```

Even though the source moved, don't we still expect factory collections to be installed? No, I guess you've taken that out of the path. Should we then test that at least the book templates folder exists? Maybe Sample Shells too? Though I suppose the program can function without samples.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-:-KKEXadQazdX676qNHz7:497788161)*
<!-- Sent from Reviewable.io -->

hatton commented:
No actual problems seem to have been found, and I needed to merge this for another team... will circle back to address the ideas for improvements.

hatton commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/BloomBrowserUI/gulpfile.js, line 23 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKE77BtmPgnqvq-xOa4:-KKKqq_FdFypOYxVe-tp:1797276411) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomBrowserUI/gulpfile.js#L23)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> If we ought to remove these (as the todo above says) why are we adding to them?
</details>
Yeah, I'm confused by the comment in the code. This prevents the jade compiler from compiling files in node_modules.

---

*[src/BloomBrowserUI/gulpfile.js, line 27 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKE7GfqwaO4Yq16u72M:-KKKr4ido4gNrD9vhAPl:1653186496) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomBrowserUI/gulpfile.js#L27)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> But surely we DO run jade and less files through compilation? Maybe there are now SOME that we need, but surely not all of them? At least the comment should change.
</details>
I read this to say, don't copy the jade, ts, less, bat, or anything in node_modules. Seems correct.

---

*[src/BloomBrowserUI/templates/template books/Decodable Reader/Basic Book.css, line 1 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKEBBuyPsqWXFIaFbqs:-KKKrKyxlTOQtNzQSfxK:1238663173) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomBrowserUI/templates/template books/Decodable Reader/Basic Book.css#L1)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> Is it intentional to check in this css file? Usually now we just check in the .less. (I assume most or all of this isn't really new so I haven't tried to review it.)
</details>
Ah, the problem is that the .gitignore spelled out a bunch of now defunct locations. I'll change it to just ignore this wherever it is found.

---

*[src/BloomExe/BloomFileLocator.cs, line 178 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKECyLeupkKLLyoeerH:-KKKwMJ3cZtzcMLrXr6X:-608283473) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomExe/BloomFileLocator.cs#L178)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> Feels odd. Why take multiple arguments and only use one? Needs comment at least.
</details>
Weird mistake. Makes no sense to have multiple parts. changed to a single "bookName".

---

*[src/BloomExe/Book/BookStorage.cs, line 590 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKESuRjE3wESXvzkHJc:-KKKx8Gb0f05afHISyCk:1790383876) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomExe/Book/BookStorage.cs#L590)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> I realize all you did here was add the html option, but the overall logic escapes me. Why is it OK to have multiple HTML files in a book folder as long as one of them is called templatePages? Why would we assume that is the book? Our factory templates are called e.g. Basic Book.htm, not templatePages.html.
</details>
Looks vestigial, to me. I don't see a templatePages anywhere.

---

*[src/BloomExe/Collection/BookCollection.cs, line 96 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKEVTSUcX52Z7LAhEGu:-KKKyJS8QdAaB6ziLEmv:872326931) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2b82016e4d5da2cf440acc5ce81b4489b0faa6b/src/BloomExe/Collection/BookCollection.cs#L96)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> Why do we want to use a name for the folder different from the name we show the user?
> 
> Should we localize any of the collection names? (Assuming this Name is displayed to the user.)
</details>
In general, I think it's simplest to keep localization in the UI layer, not down in the model. In this case, that's what happens: LibraryListView localizes this. Sure needs a comment though. Added:

// the UI and existing Localizations want to see "templates", but on disk, "templates" is ambiguous, so the name there is "template books".

---

*[src/BloomExe/MiscUI/BloomIntegrityDialog.cs, line 31 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107#-KKEWYA-wbRZLZ5f0eSG:-KKKzB9pFjBQPwckzMbi:131255472) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/eb8c9f3d3b29a842e6bc62f95e316969dc16e207/src/BloomExe/MiscUI/BloomIntegrityDialog.cs#L31)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> Even though the source moved, don't we still expect factory collections to be installed? No, I guess you've taken that out of the path. Should we then test that at least the book templates folder exists? Maybe Sample Shells too? Though I suppose the program can function without samples.
</details>
Can you think of a reason that anti-malware would remove html/css, but leave everything else?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1107)*
<!-- Sent from Reviewable.io -->
